test/CMakeLists.txt

```diff
@@ -55,56 +57,54 @@ function(add_fmt_test name)
   add_test(NAME ${name} COMMAND ${name})
 endfunction()
 
-if (FMT_MODULE)
-  return ()
-endif ()
-
-add_fmt_test(args-test)
-add_fmt_test(assert-test)
-add_fmt_test(chrono-test)
-add_fmt_test(color-test)
-add_fmt_test(core-test)
-add_fmt_test(gtest-extra-test)
-add_fmt_test(format-test mock-allocator.h)
-if (MSVC)
-  target_compile_options(format-test PRIVATE /bigobj)
-endif ()
-if (NOT (MSVC AND BUILD_SHARED_LIBS))
-  add_fmt_test(format-impl-test HEADER_ONLY header-only-test.cc)
-endif ()
-add_fmt_test(ostream-test)
-add_fmt_test(compile-test)
-add_fmt_test(compile-fp-test HEADER_ONLY)
-if (MSVC)
-  # Without this option, MSVC returns 199711L for the __cplusplus macro.
-  target_compile_options(compile-fp-test PRIVATE /Zc:__cplusplus)
-endif()
-add_fmt_test(printf-test)
-add_fmt_test(ranges-test ranges-odr-test.cc)
+if (NOT FMT_MODULE)
+  add_fmt_test(args-test)
+  add_fmt_test(assert-test)
+  add_fmt_test(chrono-test)
+  add_fmt_test(color-test)
+  add_fmt_test(core-test)
+  add_fmt_test(gtest-extra-test)
+  add_fmt_test(format-test mock-allocator.h)
+  if (MSVC)
+    target_compile_options(format-test PRIVATE /bigobj)
+  endif ()
+  if (NOT (MSVC AND BUILD_SHARED_LIBS))
+    add_fmt_test(format-impl-test HEADER_ONLY header-only-test.cc)
+  endif ()
+  add_fmt_test(ostream-test)
+  add_fmt_test(compile-test)
+  add_fmt_test(compile-fp-test HEADER_ONLY)
+  if (MSVC)
+    # Without this option, MSVC returns 199711L for the __cplusplus macro.
+    target_compile_options(compile-fp-test PRIVATE /Zc:__cplusplus)
+  endif()
+  add_fmt_test(printf-test)
+  add_fmt_test(ranges-test ranges-odr-test.cc)
 
```

Comment:
Die funktionalen Tests sollen auf jeden Fall mitlaufen. Bei Compilern, die Modules beherrschen (und nur dort!) sollen **zusätzlich** die beiden Targets `test-module` (Erstellen des BMI und der Lib) und `module-test` (Nutzen des BMI) erstellt und getestet werden.

Reply:
Klar, bisher wurde aber mit der Module version gar kein test compiliert!

Reply:
Und einfach so von der Kommandozeile? (Dateien passend zurechtgelegt)
```
clang++ -std=c++20 -fmodule-output -c -x c++ fmt.cc
clang++ -std=c++20 -I. -stdlib=libc++ test-main.cc gtest-extra.cc util.cc format.cc os.cc gmock-gtest-all.cc
clang++ -std=c++20 -I. -stdlib=libc++ -fprebuilt-module-path=. module-test.cc test-main.o gtest-extra.o util.o format.o os.o gmock-gtest-all.o fmt.o -o module-test.exe

```

Reply:
With a little more love it works:
```sh
#!/bin/sh

files="
../test/args-test.cc
../test/std-test.cc
../test/os-test.cc
../test/module-test.cc
../test/gtest-extra.cc
../test/chrono-test.cc
../test/ranges-test.cc
../test/cuda-test/cpp14.cc
../test/ostream-test.cc
../test/unicode-test.cc
../test/test-main.cc
../test/compile-test.cc
../test/scan-test.cc
../test/gtest-extra-test.cc
../test/noexception-test.cc
../test/enforce-checks-test.cc
../test/xchar-test.cc
../test/util.cc
../test/gtest/gmock-gtest-all.cc
../test/assert-test.cc
../test/ranges-odr-test.cc
../test/detect-stdfs.cc
../test/compile-fp-test.cc
../test/posix-mock-test.cc
../test/core-test.cc
../test/format-test.cc
../test/header-only-test.cc
../test/color-test.cc
../test/printf-test.cc
../test/format-impl-test.cc
../src/os.cc
../src/format.cc
../src/fmt.cc
"
for f in $files; do ln -s $f .; done

clang++ -std=c++20 -fmodule-output=fmt.pcm -I../include -c -x c++-module fmt.cc
clang++ -std=c++20 -I../include -I../test/gtest -I../test -stdlib=libc++ test-main.cc gtest-extra.cc util.cc format.cc os.cc gmock-gtest-all.cc -c
clang++ -std=c++20 -I../test -stdlib=libc++ -fprebuilt-module-path=. module-test.cc *.o -o module-test

exit

#######################
bash-3.2$ ./module-test
[==========] Running 45 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 45 tests from module_test
[ RUN      ] module_test.namespace
[       OK ] module_test.namespace (0 ms)
[ RUN      ] module_test.detail_namespace
[       OK ] module_test.detail_namespace (0 ms)
[ RUN      ] module_test.macros
[       OK ] module_test.macros (0 ms)
[ RUN      ] module_test.to_string
[       OK ] module_test.to_string (0 ms)
[ RUN      ] module_test.format
[       OK ] module_test.format (0 ms)
[ RUN      ] module_test.format_to
[       OK ] module_test.format_to (0 ms)
[ RUN      ] module_test.formatted_size
[       OK ] module_test.formatted_size (0 ms)
[ RUN      ] module_test.format_to_n
[       OK ] module_test.format_to_n (0 ms)
[ RUN      ] module_test.format_args
[       OK ] module_test.format_args (0 ms)
[ RUN      ] module_test.wformat_args
[       OK ] module_test.wformat_args (0 ms)
[ RUN      ] module_test.dynamic_format_args
[       OK ] module_test.dynamic_format_args (0 ms)
[ RUN      ] module_test.vformat
[       OK ] module_test.vformat (0 ms)
[ RUN      ] module_test.vformat_to
[       OK ] module_test.vformat_to (0 ms)
[ RUN      ] module_test.vformat_to_n
[       OK ] module_test.vformat_to_n (0 ms)
[ RUN      ] module_test.print
module-test.cc:256: Failure
fmt::print(L"{}µ", 42) produces different output.
Expected: 4\0\0\02\0\0\0?\0\0\0
  Actual: 42?
module-test.cc:257: Failure
fmt::print(__stderrp, L"{}µ", 4.2) produces different output.
Expected: 4\0\0\0.\0\0\02\0\0\0?\0\0\0
  Actual: 4.2?
[  FAILED  ] module_test.print (0 ms)
[ RUN      ] module_test.vprint
module-test.cc:265: Failure
fmt::vprint(L"{:}µ", fmt::make_wformat_args(42)) produces different output.
Expected: 4\0\0\02\0\0\0?\0\0\0
  Actual: 42?
module-test.cc:267: Failure
fmt::vprint(__stderrp, L"{}", fmt::make_wformat_args(42)) produces different output.
Expected: 4\0\0\02\0\0\0
  Actual: 42
[  FAILED  ] module_test.vprint (0 ms)
[ RUN      ] module_test.named_args
[       OK ] module_test.named_args (0 ms)
[ RUN      ] module_test.literals
[       OK ] module_test.literals (0 ms)
[ RUN      ] module_test.locale
[       OK ] module_test.locale (0 ms)
[ RUN      ] module_test.string_view
[       OK ] module_test.string_view (0 ms)
[ RUN      ] module_test.memory_buffer
[       OK ] module_test.memory_buffer (0 ms)
[ RUN      ] module_test.is_char
[       OK ] module_test.is_char (0 ms)
[ RUN      ] module_test.ptr
[       OK ] module_test.ptr (0 ms)
[ RUN      ] module_test.errors
[       OK ] module_test.errors (4 ms)
[ RUN      ] module_test.error_code
[       OK ] module_test.error_code (0 ms)
[ RUN      ] module_test.format_int
[       OK ] module_test.format_int (0 ms)
[ RUN      ] module_test.formatter
[       OK ] module_test.formatter (0 ms)
[ RUN      ] module_test.join
[       OK ] module_test.join (0 ms)
[ RUN      ] module_test.time
[       OK ] module_test.time (1 ms)
[ RUN      ] module_test.time_point
[       OK ] module_test.time_point (0 ms)
[ RUN      ] module_test.time_duration
[       OK ] module_test.time_duration (0 ms)
[ RUN      ] module_test.weekday
[       OK ] module_test.weekday (0 ms)
[ RUN      ] module_test.printf
[       OK ] module_test.printf (0 ms)
[ RUN      ] module_test.fprintf
[       OK ] module_test.fprintf (0 ms)
[ RUN      ] module_test.sprintf
[       OK ] module_test.sprintf (0 ms)
[ RUN      ] module_test.vprintf
[       OK ] module_test.vprintf (0 ms)
[ RUN      ] module_test.vfprintf
[       OK ] module_test.vfprintf (0 ms)
[ RUN      ] module_test.vsprintf
[       OK ] module_test.vsprintf (0 ms)
[ RUN      ] module_test.color
[       OK ] module_test.color (0 ms)
[ RUN      ] module_test.cstring_view
[       OK ] module_test.cstring_view (0 ms)
[ RUN      ] module_test.buffered_file
[       OK ] module_test.buffered_file (0 ms)
[ RUN      ] module_test.output_file
[=disabled=] Clang 16.0 emits multiple copies of vtables
[       OK ] module_test.output_file (0 ms)
[ RUN      ] module_test.custom_context
[       OK ] module_test.custom_context (0 ms)
[ RUN      ] module_test.compile_format_string
[=disabled=] Clang 16.0 fails to import user-defined literals
[       OK ] module_test.compile_format_string (0 ms)
[ RUN      ] module_test.std_types
[       OK ] module_test.std_types (0 ms)
[----------] 45 tests from module_test (7 ms total)

[----------] Global test environment tear-down
[==========] 45 tests from 1 test suite ran. (7 ms total)
[  PASSED  ] 43 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] module_test.print
[  FAILED  ] module_test.vprint

 2 FAILED TESTS
bash-3.2$
```

Reply:
Some progress! 🎉
The test failures are possibly related to some encoding issues. There are no functional tests for `fmt::print` with wide strings in `format-test.cc`. May be you want to add them there for a cross-check.

Reply:
Sorry, nun war ich ins Englische abgerutscht 🤦‍♀️
Ich bin nicht sicher, ob Deutsch deine Muttersprache ist. Dein Name könnte auch Schwedisch sein 😉

Reply:
Mein Name ist `Claus Klein` aus `Schwäbisch Hall` (gebürtiger Pfälzer)  ;-)